client/hotrod-client/src/test/java/org/infinispan/client/hotrod/tx/RecoveryTest.java

```diff
@@ -0,0 +1,247 @@
+      prepare(1);
+
+      timeService.advance(9000); //9 seconds, below the 10 second configured
+
+      assertBeforeTimeoutRecoveryIteration(xaResource_0, xid_0);
+      assertBeforeTimeoutRecoveryIteration(xaResource_1, xid_1);
+
+      timeService.advance(2000); //2 seconds, remote transaction will be include in recovery
+
+      assertRecoveryIteration(xaResource_0, xid_0, xid_1);
+      assertRecoveryIteration(xaResource_1, xid_1, xid_0);
+
+      //resource_1 if finished and it should be able to commit the xid_0 transaction
+      xaResource_1.commit(xid_0, false);
+      xaResource_1.rollback(xid_1);
+
+      assertEquals("v", remoteCache(0).get("k0"));
+      assertEquals(null, remoteCache(0).get("k1"));
+
+      xaResource_0.forget(xid_0);
+      xaResource_1.forget(xid_1);
+   }
+
+   protected String cacheName() {
+      return "recovery-test-cache";
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder cacheBuilder = getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, true);
+      cacheBuilder.transaction().transactionManagerLookup(new EmbeddedTransactionManagerLookup());
+      cacheBuilder.transaction().lockingMode(LockingMode.PESSIMISTIC);
+      cacheBuilder.locking().isolationLevel(IsolationLevel.REPEATABLE_READ);
+      createHotRodServers(numberOfNodes(), new ConfigurationBuilder());
+      for (EmbeddedCacheManager cm : cacheManagers) {
+         //use the same time service in all managers
+         replaceComponent(cm, TimeService.class, timeService, true);
+         //stop reaper. we are going to trigger it manually
+         extractGlobalComponent(cm, GlobalTxTable.class).stop();
+      }
+      defineInAll(cacheName(), cacheBuilder);
+   }
+
+   @Override
+   protected org.infinispan.client.hotrod.configuration.ConfigurationBuilder createHotRodClientConfigurationBuilder(
+         int serverPort) {
+      org.infinispan.client.hotrod.configuration.ConfigurationBuilder clientBuilder = super
+            .createHotRodClientConfigurationBuilder(serverPort);
+      clientBuilder.forceReturnValues(false);
+      clientBuilder.transaction().transactionManagerLookup(RemoteTransactionManagerLookup.getInstance());
+      clientBuilder.transaction().transactionMode(TransactionMode.FULL_XA);
+      clientBuilder.transaction().timeout(10, TimeUnit.SECONDS);
+      return clientBuilder;
+   }
+
+   private void assertRecoveryIteration(XAResource xaResource, Xid local, Xid remote) throws Exception {
+      Xid[] actual = xaResource.recover(XAResource.TMSTARTRSCAN);
+      assertTrue(actual.length != 0);
+      if (actual.length == 1) {
+         //it returned only the local transaction
+         assertEquals(local, actual[0]);
+         actual = xaResource.recover(XAResource.TMENDRSCAN);
+         assertEquals(1, actual.length); //other client transaction
+         assertEquals(remote, actual[0]);
+      } else {
+         //the server replied quick enough
```

Comment:
Hmmm, maybe I need to re-read that part of the code... I didn't realize recovery iteration could stop because of a timeout

Reply:
the recovery() interface supports batching. so, I'm not waiting for the server reply when it is first invoked and I'm returning whatever info the client has locally.
when the end-scan flag is used, it is where I wait for the server reply.